test/test_spectral_ops.py

```diff
@@ -43,57 +47,162 @@ def method_fn(t):
-                torch.fft.fft(t)
-            return
-
-        # CPU (MKL)
-        with self.assertRaisesRegex(RuntimeError, "MKL FFT error"):
+        with self.assertRaisesRegex(RuntimeError, match):
             torch.fft.fft(t)
+        with self.assertRaisesRegex(RuntimeError, match):
+            torch.fft.ifft(t)
+
+        with self.assertRaisesRegex(RuntimeError, match):
+            torch.fft.irfft(t)
+        with self.assertRaisesRegex(RuntimeError, match):
+            torch.fft.hfft(t)
+
+        if not t.is_complex():
+            with self.assertRaisesRegex(RuntimeError, match):
+                torch.fft.rfft(t)
+            with self.assertRaisesRegex(RuntimeError, match):
+                torch.fft.ihfft(t)
+
+    def test_fft_invalid_dtypes(self, device):
+        t = torch.randn(64, device=device, dtype=torch.complex128)
+
+        with self.assertRaisesRegex(RuntimeError, "Expected a real input tensor"):
+            torch.fft.rfft(t)
+
+        with self.assertRaisesRegex(RuntimeError, "Expected a real input tensor"):
+            torch.fft.ihfft(t)
 
-    @dtypes(torch.int64, torch.float32)
-    def test_fft_invalid_dtypes(self, device, dtype):
-        if dtype.is_floating_point:
+    @skipCUDAIfRocm
+    @skipCPUIfNoMkl
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.int8, torch.float, torch.double, torch.complex64, torch.complex128)
+    def test_fft_type_promotion(self, device, dtype):
+        if dtype.is_complex or dtype.is_floating_point:
             t = torch.randn(64, device=device, dtype=dtype)
         else:
             t = torch.randint(-2, 2, (64,), device=device, dtype=dtype)
 
-        with self.assertRaisesRegex(RuntimeError, "Expected a complex tensor"):
-            torch.fft.fft(t)
+        PROMOTION_MAP = {
+            torch.int8: torch.complex64,
+            torch.float: torch.complex64,
+            torch.double: torch.complex128,
+            torch.complex64: torch.complex64,
+            torch.complex128: torch.complex128,
+        }
+        T = torch.fft.fft(t)
+        self.assertEqual(T.dtype, PROMOTION_MAP[dtype])
+
+        PROMOTION_MAP_C2R = {
+            torch.int8: torch.float,
+            torch.float: torch.float,
+            torch.double: torch.double,
+            torch.complex64: torch.float,
+            torch.complex128: torch.double,
+        }
+        R = torch.fft.hfft(t)
+        self.assertEqual(R.dtype, PROMOTION_MAP_C2R[dtype])
+
+        if not dtype.is_complex:
+            PROMOTION_MAP_R2C = {
+                torch.int8: torch.complex64,
+                torch.float: torch.complex64,
+                torch.double: torch.complex128,
+            }
+            C = torch.fft.rfft(t)
+            self.assertEqual(C.dtype, PROMOTION_MAP_R2C[dtype])
+
+    @skipCPUIfNoMkl
+    @skipCUDAIfRocm
+    @dtypes(torch.double, torch.complex128)  # gradcheck requires double
+    def test_fft_backward(self, device, dtype):
```

Comment:
This is cool and I like checking gradients directly in these "vertical" test suites. Maybe we should still add method_test entries for the moment?